man/systemd-sysupdate.xml

```diff
@@ -0,0 +1,284 @@
+        implicitly invoked whenever a new update is initiated.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><option>pending</option></term>
+
+        <listitem><para>Tests whether a newer version of the OS is installed than the one currently
+        running. Returns zero if so, non-zero otherwise. This compares the newest installed version's
+        identifier with the OS image version as reported by the <varname>IMAGE_VERSION=</varname> field in
+        <filename>/etc/os-release</filename>. If the former is newer than the latter, an update was
+        apparently completed but not activated (i.e. rebooted into) yet.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><option>reboot</option></term>
+
+        <listitem><para>Similar to the <option>pending</option> command but immediately reboots in case a
+        newer version of the OS has been installed than the one currently running. This operation can be done
+        implicitly together with the <command>now</command> command, after a completed update via the
+        <option>--reboot</option> switch, see below.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><option>components</option></term>
+
+        <listitem><para>Lists components that can be updated. This enumerates the
+        <filename>/etc/sysupdate.*.d/</filename>, <filename>/run/sysupdate.*.d/</filename> and
+        <filename>/usr/lib/sysupdate.*.d/</filename> directories that contain transfer files. This command is
+        useful to list possible parameters for <option>--component=</option> (see below).</para></listitem>
+      </varlistentry>
+
+      <xi:include href="standard-options.xml" xpointer="help" />
+      <xi:include href="standard-options.xml" xpointer="version" />
+    </variablelist>
+  </refsect1>
+
+
+  <refsect1>
+    <title>Options</title>
+
+    <para>The following options are understood:</para>
+
+    <variablelist>
+
+      <varlistentry>
+        <term><option>--component=</option></term>
+        <term><option>-C</option></term>
+
+        <listitem><para>Selects the component to update. Takes a component name as argument. This has the
+        effect of slightly altering the search logic for transfer files. If this switch is not used, the
+        transfer files are loaded from <filename>/etc/sysupdate.d/*.conf</filename>,
+        <filename>/run/sysupdate.d/*.conf</filename> and <filename>/usr/lib/sysupdate.d/*.conf</filename>. If
+        this switch is used, the specified component name is used to alter the directories to look in to be
+        <filename>/etc/sysupdate.&lt;component&gt;.d/*.conf</filename>,
+        <filename>/run/sysupdate.&lt;component&gt;.d/*.conf</filename> and
+        <filename>/usr/lib/sysupdate.&lt;component&gt;.d/*.conf</filename>, each time with the
+        <filename>&lt;component&gt;</filename> string replaced with the specified component name.</para>
```

Comment:
I think this convention of using separate directories is going to make the whole thing harder to use than it should. For example, `systemd-analyze cat-config` cannot show this, and would need to have special logic. But even when scripting, dealing with multiple directories is annoying.

I think this grouping should be performed by using `sysupdate.d/<component>.<part>.conf`. Then all files with `<component>.` prefix can be understood to be one component, and files without a dot are not part of a component. This is easier to list and process.

An additional advantage is that with a "flat" file list, we use the same sorting and replacement logic as in other components. So for example we know that `aaa.conf` is before `component.b.conf` and before `c-11.conf`. With directory-components, we'd need to figure out a separate sorting convention.

And finally, a separate `--component` and `--definitions` parts are not needed. Just specify `--component=aaa` or `--component=component.*` and it all works.

Reply:
I actually started out from that idea, i.e. instead of doing one dir per component, encode the component in the filename and allow them to be placed in the same dir. But I figured this would be more complex in the end, and not match how we did this elsewhere in drop-in dirs, in particular in repart.d/ drop-ins.

i.e. i figured it would be simplest to say: have one dir where we drop files in. i.e. at the most basic the logic that "directory" means grouping and "regular file" in it is the stuff that is grouped, is such an extremely basic idea that it i wanted to focus on it. --definitions= allows you to specify such a directory and read the configs from it.

But then I realized that it is useful to also enforce a naming regime, i.e. there should be an easy way to enumerate these directories with those files in them. So the next idea i had was to name them `/usr/lib/sysupdate.d/<component>.d/<transfer>.conf`. But I didn't like that because it put so much emphasis on the component concept, which is a concept I didn#t want so much emphasis on. It should be a concept you shouldn't have ot know about and i'd guess in a major part of the usecases for sysupdate you'll only ever have the OS itself to update, but no other components.

That's why I came to this idea that `/usr/lib/sysupdate.d/<transfer>.conf` is the where the main component is, and `/usr/lib/sysupdate.<component>.d/<transfer>.conf` can be used to define additional components.

The net effect of that is that noone has to think about components, or be aware of them. As long as you just want to update your OS, you can entirely ignore them. And that's a *good* thing.

Reply:
I'd expect 90% of cases will just update the system with this. (Or they update nspawn or portable images with it, but in that case the components concept doesn't matter, as you#d just embed the sysupdate dropins inside of the image, so that there will only be one component for each image.).